File: api/app/services/rules.py
from typing import Any


def evaluate_show_if(operator: str, actual: Any, trigger_value: Any) -> bool:
    if operator == "eq":
        return actual == trigger_value
    if operator == "neq":
        return actual != trigger_value
    if operator == "in":
        return isinstance(trigger_value, list) and actual in trigger_value
    if operator == "not_in":
        return isinstance(trigger_value, list) and actual not in trigger_value
    return False


def item_is_visible(rules: list[dict[str, Any]], answers: dict[str, Any]) -> bool:
    if not rules:
        return True
    for rule in rules:
        if rule.get("type") != "show_if":
            continue
        code = rule.get("trigger_question_code")
        actual = answers.get(code)
        if not evaluate_show_if(rule.get("operator", ""), actual, rule.get("trigger_value")):
            return False
    return True

File: packages/shared/src/rules.ts
import type { AnswersMap, Item, ShowIfRule } from "./types";

function isMissing(value: unknown): boolean {
  return value === undefined || value === null;
}

function isPrimitive(value: unknown): value is string | number | boolean | null {
  return (
    value === null ||
    typeof value === "string" ||
    typeof value === "number" ||
    typeof value === "boolean"
  );
}

// Deterministic, fail-closed evaluation:
// - missing answer => false
// - unknown operator => false
// - type mismatch => false (no coercion)
export function isRuleMatch(rule: ShowIfRule, answers: AnswersMap): boolean {
  const actual = answers[rule.trigger_question_code];
  if (isMissing(actual)) return false;

  switch (rule.operator) {
    case "eq": {
      if (!isPrimitive(actual) || Array.isArray(rule.trigger_value)) return false;
      return actual === rule.trigger_value;
    }
    case "neq": {
      if (!isPrimitive(actual) || Array.isArray(rule.trigger_value)) return false;
      return actual !== rule.trigger_value;
    }
    case "in": {
      if (!isPrimitive(actual) || !Array.isArray(rule.trigger_value)) return false;
      return rule.trigger_value.includes(actual);
    }
    case "not_in": {
      if (!isPrimitive(actual) || !Array.isArray(rule.trigger_value)) return false;
      return !rule.trigger_value.includes(actual);
    }
    default:
      return false;
  }
}

export function evaluateShowIf(rules: ShowIfRule[] | undefined, answers: AnswersMap): boolean {
  if (!rules || rules.length === 0) return true;
  return rules.every((rule) => isRuleMatch(rule, answers));
}

export function isItemVisible(item: Item, answers: AnswersMap): boolean {
  return evaluateShowIf(item.rules, answers);
}
